OpenRA.Mods.Common/Activities/Attack.cs

```diff
@@ -72,27 +72,63 @@ public override Activity Tick(Actor self)
 			return NextActivity;
 		}
 
+		protected bool ValidateFrozenActorTarget(Actor self, AttackBase attack)
+		{
+			// If target was a frozen actor got replaced with an actor, try switching
+			if (!target.IsValidFor(self))
+			{
+				if (target.FrozenActor != null && target.FrozenActor.Actor != null)
+				{
+					var loc = self.World.Map.CellContaining(target.FrozenActor.CenterPosition);
+					foreach (var actor in self.World.ActorMap.GetActorsAt(loc))
+					{
+						if (actor.ActorID == target.FrozenActor.ID && self.Owner.CanTargetActor(actor))
+						{
+							target = Target.FromActor(actor);
+							return true;
+						}
+					}
+				}
+
+				return false;
+			}
+
+			var positionableLoc = self.World.Map.CellContaining(target.CenterPosition);
+			if (attack.Info.AttackRequiresEnteringCell && !positionable.CanEnterCell(positionableLoc, null, false))
+				return false;
+
+			// If target was an actor that got replaced with a frozen actor, try switching
+			if (target.Actor != null && !self.Owner.CanTargetActor(target.Actor))
+			{
+				if (target.Type == TargetType.Actor)
+				{
+					var frozenLayer = self.Owner.PlayerActor.TraitOrDefault<FrozenActorLayer>();
+					var frozenActor = frozenLayer.FromID(target.Actor.ActorID);
+					if (frozenActor != null && target.IsValidFor(self))
+					{
+						target = Target.FromFrozenActor(frozenActor);
+						return true;
+					}
+				}
+
+				return false;
+			}
+
+			return true;
+		}
+
 		protected virtual Activity InnerTick(Actor self, AttackBase attack)
 		{
 			if (IsCanceled)
 				return NextActivity;
 
-			var type = Target.Type;
-			if (!Target.IsValidFor(self) || type == TargetType.FrozenActor)
-				return NextActivity;
-
-			if (attack.Info.AttackRequiresEnteringCell && !positionable.CanEnterCell(Target.Actor.Location, null, false))
-				return NextActivity;
+			var type = target.Type;
 
-			// Drop the target if it moves under the shroud / fog.
-			// HACK: This would otherwise break targeting frozen actors
-			// The problem is that Shroud.IsTargetable returns false (as it should) for
-			// frozen actors, but we do want to explicitly target the underlying actor here.
-			if (!attack.Info.IgnoresVisibility && type == TargetType.Actor && !Target.Actor.Info.HasTraitInfo<FrozenUnderFogInfo>() && !self.Owner.CanTargetActor(Target.Actor))
+			if (!ValidateFrozenActorTarget(self, attack))
```

Comment:
> Do we ever have a case where target.IsValidFor(self) == true but self.Owner.CanTargetActor(target.Actor) == false 

I am not sure. I suspect that this is not the case, but I do not know how to check for sure. From looking at the original code it seems like it was not written under the assumption that this is possibe, since the second check is part of the shroud hack while the first was combined with `|| type == frozen`. I am going to try to assume that case yoou described does not happen.